python/mspasspy/db/database.py

```diff
@@ -574,45 +645,182 @@ def save_data(self, mspass_object, mode="promiscuous", storage_mode='gridfs', di
-                    update_dict['dfile'] = dfile
-                    update_dict['foff'] = foff
-                    if format:
-                        update_dict['nbytes'] = nbytes
-                        update_dict['format'] = format
-                elif storage_mode == "gridfs":
-                    old_gridfs_id = None if 'gridfs_id' not in object_doc else object_doc['gridfs_id']
-                    gridfs_id = self._save_data_to_gridfs(mspass_object, old_gridfs_id)
-                    update_dict['gridfs_id'] = gridfs_id
+            insertion_dict['storage_mode']=storage_mode
+
+            if storage_mode == "file":
+                # TODO:  be sure this can't throw an exception
+                foff, nbytes = self._save_data_to_dfile(mspass_object, dir, dfile, format=format)
+                insertion_dict['dir'] = dir
+                insertion_dict['dfile'] = dfile
+                insertion_dict['foff'] = foff
+                if format:
+                    insertion_dict['nbytes'] = nbytes
+                    insertion_dict['format'] = format
+            elif storage_mode == "gridfs":
+                if overwrite and 'gridfs_id' in insertion_dict:
+                    gridfs_id = self._save_data_to_gridfs(mspass_object, insertion_dict['gridfs_id'])
+                else:
+                    gridfs_id = self._save_data_to_gridfs(mspass_object)
+                insertion_dict['gridfs_id'] = gridfs_id
                 #TODO will support url mode later
                 #elif storage_mode == "url":
                 #    pass
-                col.update_one(filter_, {'$set': update_dict})
+                #col.update_one(filter_, {'$set': update_dict})
+
+            if mspass_object.is_empty():
+                history_object_id = None
+            else:
+                # optional history save - only done if history container is not empty
+                history_obj_id_name = self.database_schema.default_name('history_object') + '_id'
+                history_object_id = self._save_history(mspass_object, alg_name, alg_id)
+                insertion_dict[history_obj_id_name] = history_object_id
+
+            if data_tag:
+                insertion_dict['data_tag']=data_tag
+            else:
+                # We need to clear data tag if was previously defined in
+                # this case or a the old tag will be saved with this datum
+                if 'data_tag' in insertion_dict:
+                    insertion_dict.erase('data_tag')
+            # finally ready to insert the wf doc - keep the id as we'll need
+            # it for tagging any elog entries
+            wfid = wf_collection.insert_one(insertion_dict).inserted_id
+            # Put wfid into the object's meta as the new definition of 
+            # the parent of this waveform
+            # TODO - I think we need to reset the history container here 
+            # and mark this new save as an ORIGIN leaving till later
```

Comment:
This is done by the _save_history method in database.py.

Reply:
Ok, I removed that comment then.  Did not see that - thanks